File: 1161-maximum-level-sum-of-a-binary-tree/1161-maximum-level-sum-of-a-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxLevelSum(root *TreeNode) int {
    maxSum, ans, level := int(-1 << 31), 0, 0

    queue := []*TreeNode{root}

    for len(queue) > 0 {
        level++
        sumPerLevel := 0
        levelSize := len(queue)
        for i := 0; i < levelSize; i++ {
            node := queue[0]
            queue = queue[1:] // pop

            sumPerLevel += node.Val

            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }

        if sumPerLevel > maxSum {
            maxSum = sumPerLevel
            ans = level
        }
    }

    return ans
}

// bfs, binary tree, tree
// time: O(n)
// space: O(n)

File: 1161-maximum-level-sum-of-a-binary-tree/1161-maximum-level-sum-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        max_sum, ans, level = -100000, 0, 0
        q = deque([root])
        while q:
            level += 1
            sum_per_level = 0
            level_size = len(q)
            for _ in range(level_size):
                curr = q.popleft()
                sum_per_level += curr.val
                
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)

            if sum_per_level > max_sum:
                max_sum = sum_per_level
                ans = level
        
        return ans

# bfs, binary tree, tree
# time: O(n)
# space: O(n)